core/drivers/stm32_etzpc.c

```diff
@@ -66,263 +61,400 @@
+		if (pdata->tzma_cfg[n] & TZMA_PM_LOCK_BIT)
+			etzpc_do_lock_tzma(etzpc_dev, n);
 	}
 
 	return TEE_SUCCESS;
 }
 DECLARE_KEEP_PAGER(etzpc_pm);
 
-static void init_pm(struct etzpc_instance *dev)
+static struct etzpc_device *stm32_etzpc_alloc(void)
 {
-	unsigned int n = 0;
-
-	dev->periph_cfg = calloc(dev->num_per_sec, sizeof(*dev->periph_cfg));
-	dev->tzma_cfg = calloc(dev->num_tzma, sizeof(*dev->tzma_cfg));
-	if (!dev->periph_cfg || !dev->tzma_cfg)
-		panic();
+	struct etzpc_device *etzpc_dev = calloc(1, sizeof(*etzpc_dev));
+	struct stm32_etzpc_driver_data *ddata = calloc(1, sizeof(*ddata));
 
-	for (n = 0; n < dev->num_per_sec; n++) {
-		dev->periph_cfg[n] = (uint8_t)etzpc_get_decprot(n);
-		if (etzpc_get_lock_decprot(n))
-			dev->periph_cfg[n] |= PERIPH_PM_LOCK_BIT;
+	if (etzpc_dev && ddata) {
+		etzpc_dev->ddata = ddata;
+		return etzpc_dev;
 	}
 
-	for (n = 0; n < dev->num_tzma; n++) {
-		dev->tzma_cfg[n] = (uint8_t)etzpc_get_tzma(n);
-		if (etzpc_get_lock_tzma(n))
-			dev->tzma_cfg[n] |= TZMA_PM_LOCK_BIT;
-	}
+	free(ddata);
+	free(etzpc_dev);
 
-	register_pm_core_service_cb(etzpc_pm, dev, "stm32-etzpc");
+	return NULL;
 }
 
-struct etzpc_hwcfg {
-	unsigned int num_tzma;
-	unsigned int num_per_sec;
-	unsigned int num_ahb_sec;
-	unsigned int chunk_size;
-};
+/* Informative unused function */
+static __unused void stm32_etzpc_free(struct etzpc_device *etzpc_dev)
+{
+	if (etzpc_dev) {
+		free(etzpc_dev->ddata);
+		free(etzpc_dev);
+	}
+}
 
-static void get_hwcfg(struct etzpc_hwcfg *hwcfg)
+static void stm32_etzpc_set_driverdata(struct etzpc_device *dev)
 {
-	uint32_t reg = io_read32(etzpc_base() + ETZPC_HWCFGR);
+	struct stm32_etzpc_driver_data *ddata = dev->ddata;
+	vaddr_t base = dev->pdata.base.va;
+	uint32_t reg = io_read32(base + ETZPC_HWCFGR);
 
-	hwcfg->num_tzma = (reg & ETZPC_HWCFGR_NUM_TZMA_MASK) >>
+	ddata->num_tzma = (reg & ETZPC_HWCFGR_NUM_TZMA_MASK) >>
 			  ETZPC_HWCFGR_NUM_TZMA_SHIFT;
-	hwcfg->num_per_sec = (reg & ETZPC_HWCFGR_NUM_PER_SEC_MASK) >>
+	ddata->num_per_sec = (reg & ETZPC_HWCFGR_NUM_PER_SEC_MASK) >>
 			     ETZPC_HWCFGR_NUM_PER_SEC_SHIFT;
-	hwcfg->num_ahb_sec = (reg & ETZPC_HWCFGR_NUM_AHB_SEC_MASK) >>
+	ddata->num_ahb_sec = (reg & ETZPC_HWCFGR_NUM_AHB_SEC_MASK) >>
 			     ETZPC_HWCFGR_NUM_AHB_SEC_SHIFT;
-	hwcfg->chunk_size = (reg & ETZPC_HWCFGR_CHUNCKS1N4_MASK) >>
-			    ETZPC_HWCFGR_CHUNCKS1N4_SHIFT;
+
+	DMSG("ETZPC revision 0x%02"PRIx8", per_sec %u, ahb_sec %u, tzma %u",
+	     io_read8(base + ETZPC_VERR),
+	     ddata->num_per_sec, ddata->num_ahb_sec, ddata->num_tzma);
 }
 
-static void init_device_from_hw_config(struct etzpc_instance *dev,
-				       paddr_t pbase)
+static void fdt_etzpc_conf_decprot(struct etzpc_device *dev,
+				   const void *fdt, int node)
 {
-	struct etzpc_hwcfg hwcfg = { };
+	const fdt32_t *cuint = NULL;
+	size_t i = 0;
+	int len = 0;
+
+	cuint = fdt_getprop(fdt, node, "default-access-conf", &len);
+	if (!cuint) {
+		DMSG("No ETZPC DECPROT configuration in DT");
+		return;
+	}
```

Comment:
maybe `if (len & 1) panic("Invalid ETZPC configuration");`

Reply:
Reverted this change